Add CREATE2factory

AugustoL opened the pull request:
Fixes #1644 

This PR adds a simple CREATE2Factory based on different implementations I saw around.
The goal is to have the minimum and necessary contract factory to deploy contracts using CREATE2 evm opcode. This I why I didnt added any extra methods to deploy smart contracts with a fixed bytecode, nevertheless this contract can be extended into metatxs or ownable create2 factories.

AugustoL commented:
@spalladino I followed the same nomeclature used in https://github.com/zeppelinos/zos/blob/master/packages/lib/contracts/upgradeability/ProxyFactory.sol

I dont know if you plan to allow CREATE2 to be used in the proxy factory, but I guess thats something that should be discussed in zeppelinos repo.

ping to @nventuro @cwhinfrey who discussed the implementation in https://github.com/OpenZeppelin/openzeppelin-solidity/issues/1644

I Think this PR should be focused on a simple implementation of a CREATE2 factory. And once accepted we can extend this to two more factories, one that support Ownable contracts and another that deploys contracts with a fixed bytecode, where the fixed bytecode version can be used for metatxs contracts.

nventuro commented on contracts/utils/CREATE2Factory.sol:
```diff
@@ -0,0 +1,64 @@
+    ) public view returns (address) {
+        bytes32 codeHash = keccak256(code);
+        bytes32 _data = keccak256(
+            abi.encodePacked(bytes1(0xff), deployer, salt, codeHash)
```
I'm confused, shouldn't `deployer` be `address(this)`?

cwhinfrey replied:
@nventuro Agreed it should be `address(this)`

AugustoL replied:
yes, my bad. changing it

nventuro commented on contracts/utils/CREATE2Factory.sol:
```diff
@@ -0,0 +1,64 @@
+     */
+    function _deploy(
+        bytes32 _salt, bytes memory _code
+    ) internal returns(address) {
```
```suggestion
    ) private returns(address) {
```

No need to expose low level details outside of this contract.

nventuro commented on contracts/utils/CREATE2Factory.sol:
```diff
@@ -0,0 +1,64 @@
+ * where the smart contracts with the specified salt and bytecode would be
+ * deployed.
+ */
+contract CREATE2Factory {
```
I think we could simply call this `Create2`, and we'd end up with functions like `Create2.deploy`, etc. Not a fan of exposing such a low level detail in the name, but it's succinct, and everyone already knows about it. I'm open to alternatives.

AugustoL replied:
Couldn't agree more. :+1:

nventuro commented on contracts/utils/CREATE2Factory.sol:
```diff
@@ -0,0 +1,64 @@
+        }
+        return _addr;
+    }
+
```
```suggestion
```

Giulio2002 commented:
After having looked at this PR, I would like to suggest to split CREATE2Factory into two contracts. a BasicFactory Contract without public functions and then a standard factory Contract which contains public functions such as deploy and computeAddress. so that if someone wants to create a Factory that takes different arguments in deploy and computeAddress we can still make it inherit from BasicFactory without completely rewriting the contract. an example of this kind of scenario has been provided by @marekkirejczyk in #1651.

cwhinfrey commented on contracts/utils/CREATE2Factory.sol:
```diff
@@ -0,0 +1,64 @@
+     */
+    function computeAddress(
+        address deployer, bytes32 salt, bytes memory code
+    ) public view returns (address) {
```
I think this can be restricted to `pure`

cwhinfrey commented:
@AugustoL this is awesome! Great to see this coming along. I think one thing to think about is whether people will want to use the regular constructor or the `initialize` pattern like @spalladino is using in the [ZOS Create2 Factory](https://github.com/zeppelinos/zos/blob/master/packages/lib/contracts/upgradeability/ProxyFactory.sol#L58) and whether or not we should be opinionated about it here.

Using the regular constructor will mean appending any constructor parameters to the end of the init code before passing it in as the `code` parameter in `deploy()`. It will also mean that constructor parameters are locked with the reserved address because changing them will change the address. That's probably desirable for some use cases and may not be for others.

Using the `initialize` pattern is a bit more flexible in terms of whether the constructor parameters affect the address the contract will be deployed to. However, providing direct support for it may be too opinionated for this depending on how it's done.

After thinking through all of this, I really like @nventuro's idea of having a Create2 library rather than a factory (or having a library _and_ a factory that uses it). It would give people the most flexibility in how they calculate the salt, initialize the contracts, ect. while still encapsulating the Create2 assembly code and address calculation.

nventuro commented:
@cwhinfrey a flexible library that simply makes the opcode easier to use and then building on top of that with common patterns sounds like the way to go :)

AugustoL commented:
@nventuro I changed the contract for a library. Next:

- We have 2 computeAddress function, one that can receive the deployer address and one that takes in count address(this) as deployer. Should we leave it like that? I think is good to have both of them, other contracts using this libs might want to compute address with a deployer address.
- I added an example of Create2 for Ownable contracts, but if you wnat we can remove it, I think is good to have an example of how to use the library.
- For initilizable contracts we can add a `deployAndInitilize` funciton in the library or an example of how to use it with Initilizable contracts, I think having a funciton in the library would be better. WDYT? @cwhinfrey @spalladino

nventuro commented:
>We have 2 computeAddress function, one that can receive the deployer address and one that takes in count address(this) as deployer. Should we leave it like that? I think is good to have both of them, other contracts using this libs might want to compute address with a deployer address.

I like this, since a user may either use this to be a factory itself, or to compute addresses of contracts deployed by other factories. I worry that the names being the same may make this confusing, but am unsure as to what a better name would be (`computeOwnDeployedAddress`? ew).

The `Ownable` example is interesting, thanks! We may want to include such a thing in a guide.

Note that the `library` functions are `public`: we should make them `internal` so that contracts can use it by having the code inlined (as opposed to needing to deploy a separate library contract).

AugustoL commented:
@nventuro what about having only the `computeAddress` function with the deployer arg? Most of the cases it will be used with address(this) but not all the time, it is more flexible and we wont have "duplicated" code.

Also, what about having an anotrher example of a create2 factory for meta txs? that can be useful.

nventuro commented:
>what about having only the computeAddress function with the deployer arg?

I was also thinking that having just the more flexible one sounds like the way to go, but now I wonder, in which scenario would you need to compute the address where _another_ factory would deploy a contract? @spalladino thoughts?

nventuro commented:
We'd really like to have support for `create2` in some way in OpenZeppelin. If anyone has specific thoughts regarding capabilities, API, etc., please share here!

marekkirejczyk commented:
A couple of comments after working with @JustynaBroniszewska on CREATE2 based functionality in Universal Login:

* `Create2.deploy` could return address, so that in example `Create2OwnableFactory` we don't need to calculate it
* Emitting events seems unnecessary in some cases hence always cost gas, I would propose to move it to user's Factory code.
* It seems `computeAddress` don't need to be implemented in the smart contract, could be js function, that would look sth like this:
```ts 
const computeAddress = (deployer, salt, code) => {
  const codeHash = keccak256(['bytes'], [code]);
  const hashedData = keccak256(['bytes1', 'address', 'bytes32', 'bytes32'], ['0xff', deployer, salt, codeHash]);
  return `0x${hashedData.slice(26)}`;
};
```

Looking forward to your opinions.

ngmachado commented:
> A couple of comments after working with @JustynaBroniszewska on CREATE2 based functionality in Universal Login:
> 
> * `Create2.deploy` could return address, so that in example `Create2OwnableFactory` we don't need to calculate it
> * Emitting events seems unnecessary in most cases, but always cost gas, I would propose to move it to user's Factory code.
> * It seems `computeAddress` don't need to be implemented in the smart contract, could be js function, that would look sth like this:
> 
> ```ts
> const computeAddress = (deployer, salt, code) => {
>   const codeHash = keccak256(['bytes'], [code]);
>   const hashedData = keccak256(['bytes1', 'address', 'bytes32', 'bytes32'], ['0xff', deployer, salt, codeHash]);
>   return `0x${hashedData.slice(26)}`;
> };
> ```
> 
> Looking forward to your opinions.

Hi, 

The two firsts point seems ok, but the third, I think the lib should maintain a method to calculate the address by itself. (is kinda a syntax sugar)

We can build some smart contract logic by calculating the address without the need of external source (feed by the user).  

Cheers

ngmachado commented:
> > A couple of comments after working with @JustynaBroniszewska on CREATE2 based functionality in Universal Login:
> > 
> > * `Create2.deploy` could return address, so that in example `Create2OwnableFactory` we don't need to calculate it
> > * Emitting events seems unnecessary in most cases, but always cost gas, I would propose to move it to user's Factory code.
> > * It seems `computeAddress` don't need to be implemented in the smart contract, could be js function, that would look sth like this:
> > 
> > ```ts
> > const computeAddress = (deployer, salt, code) => {
> >   const codeHash = keccak256(['bytes'], [code]);
> >   const hashedData = keccak256(['bytes1', 'address', 'bytes32', 'bytes32'], ['0xff', deployer, salt, codeHash]);
> >   return `0x${hashedData.slice(26)}`;
> > };
> > ```
> > 
> > 
> > Looking forward to your opinions.
> 
> Hi,
> 
> The two firsts point seems ok, but the third, I think the lib should maintain a method to calculate the address by itself. (is kinda a syntax sugar)
> 
> We can build some smart contract logic by calculating the address without the need of external source (feed by the user).
> 
> Cheers

I found this PR because i was looking for that specific method 💯

marekkirejczyk commented:
> > > A couple of comments after working with @JustynaBroniszewska on CREATE2 based functionality in Universal Login:
> > > 
> > > * `Create2.deploy` could return address, so that in example `Create2OwnableFactory` we don't need to calculate it
> > > * Emitting events seems unnecessary in most cases, but always cost gas, I would propose to move it to user's Factory code.
> > > * It seems `computeAddress` don't need to be implemented in the smart contract, could be js function, that would look sth like this:
> > > 
> > > ```ts
> > > const computeAddress = (deployer, salt, code) => {
> > >   const codeHash = keccak256(['bytes'], [code]);
> > >   const hashedData = keccak256(['bytes1', 'address', 'bytes32', 'bytes32'], ['0xff', deployer, salt, codeHash]);
> > >   return `0x${hashedData.slice(26)}`;
> > > };
> > > ```
> > > 
> > > 
> > > Looking forward to your opinions.
> > 
> > 
> > Hi,
> > The two firsts point seems ok, but the third, I think the lib should maintain a method to calculate the address by itself. (is kinda a syntax sugar)
> > We can build some smart contract logic by calculating the address without the need of external source (feed by the user).
> > Cheers
> 
> I found this PR because i was looking for that specific method 💯

Sounds fair. 
I propose adding a js util function, as it would come in handy for some cases.

ngmachado commented:
Hi everybody,

It looks like the is a code duplication in two methods, why not encapsulate as :

```solidity
 function computeAddress(bytes32 salt, bytes memory codeInit) internal returns(address) {
        return computeAddress(address(this), salt, codeInit);
 }
```

The deploy function should return an address instead of emitting an event as:

```solidity
function deploy(bytes32 salt, bytes memory codeInit) internal returns(address) {
        return _deploy(salt, codeInit);
}
```


Cheers

Skyge commented on contracts/utils/Create2.sol:
```diff
@@ -0,0 +1,79 @@
+     * @param code The bytecode of of the contract to be deployed
+     */
+    function deploy(bytes32 salt, bytes memory code) internal {
+        address addr = _deploy(salt, code);
```
Maybe you can use `deployAddress` to replace `addr`, it is shorter, but it will be more difficult to understand.

Skyge commented:
@ngmachado 

> It looks like the is a code duplication in two methods, why not encapsulate as :

Not really, just like the function name, one is to deploy, and the other is to calculate. you can use `computeAddress()` to find an expected contract address with a suitable `salt`, and then you can use `deploy()` to get that expected address.

Skyge commented:
> > A couple of comments after working with @JustynaBroniszewska on CREATE2 based functionality in Universal Login:
> > 
> > * `Create2.deploy` could return address, so that in example `Create2OwnableFactory` we don't need to calculate it
> > * Emitting events seems unnecessary in most cases, but always cost gas, I would propose to move it to user's Factory code.
> > * It seems `computeAddress` don't need to be implemented in the smart contract, could be js function, that would look sth like this:
> > 
> > ```ts
> > const computeAddress = (deployer, salt, code) => {
> >   const codeHash = keccak256(['bytes'], [code]);
> >   const hashedData = keccak256(['bytes1', 'address', 'bytes32', 'bytes32'], ['0xff', deployer, salt, codeHash]);
> >   return `0x${hashedData.slice(26)}`;
> > };
> > ```
> > 
> > 
> > Looking forward to your opinions.
> 
> Hi,
> 
> The two firsts point seems ok, but the third, I think the lib should maintain a method to calculate the address by itself. (is kinda a syntax sugar)
> 
> We can build some smart contract logic by calculating the address without the need of external source (feed by the user).
> 
> Cheers

We can keep both,  one more approach, one more choice.

Skyge commented on contracts/examples/Create2OwnableFactory.sol:
```diff
@@ -0,0 +1,17 @@
+import "../ownership/Ownable.sol";
+
+contract Create2OwnableFactory {
+  function deploy(bytes32 salt, bytes memory code) public {
```
If I use `uint256` rather than `bytes32`, is it OK?

cwhinfrey replied:
@Skyge I believe it will be fairly common to calculate the salt with a `keccak256` hash. Is there a case where the salt would be calculated as an integer?

ngmachado commented:
> @ngmachado
> 
> > It looks like the is a code duplication in two methods, why not encapsulate as :
> 
> Not really, just like the function name, one is to deploy, and the other is to calculate. you can use `computeAddress()` to find an expected contract address with a suitable `salt`, and then you can use `deploy()` to get that expected address.

I was thinking in maintaining the two methods, just make one call the other if you dont pass the first param (deployer address). The code inside the two function are same.

Skyge commented on contracts/examples/Create2OwnableFactory.sol:
```diff
@@ -0,0 +1,17 @@
+  ) public view returns (address) {
+      return Create2.computeAddress(salt, code);
+  }
+}
```
In the contract `Create2.sol`, you have two approach named `computeAddress()`, so is there a necessary to add the other function 
```javascript
function computeAddress(
      address deployer, bytes32 salt, bytes memory code
  ) public view returns (address) {
      return Create2.computeAddress(deployer, salt, code);
  }
```

Skyge commented:
@gmachado Ohhhh, I am sorry, I misunderstood what you meant. 
As for the first, two methods have different arguments, so maybe it is not suitable to encapsulate. 
As for the second, yeah, I think he @AugustoL  should put the event `emit Create2Deployed(addr, salt);` into the the **private** function like this
```javascript
    function _deploy(
        bytes32 _salt, bytes memory _code
    ) private returns(address) {
        address _addr;
        // solhint-disable-next-line no-inline-assembly
        assembly {
            _addr := create2(0, add(_code, 0x20), mload(_code), _salt)
            if iszero(extcodesize(_addr)) {
                revert(0, 0)
            }
        }
        emit Create2Deployed(_addr, _salt);
        return _addr;
    }
```

Skyge commented:
Here is another [Create2Factory contract](https://etherscan.io/address/0x000000000063b99b8036c31e91c64fc89bff9ca7#contracts), you can see 

 > The contract prevents redeploys by maintaining a mapping of all
 contracts that have already been deployed, and prevents frontrunning or other
 collisions by requiring that the first 20 bytes of the salt are equal to the
  address of the caller (this can be bypassed by setting the first 20 bytes to
 the null address). There is also a view function that computes the address of
 the contract that will be created when submitting a given salt or nonce along
 with a given block of initialization code.

it mentions frontrunning. And also there is a repo named [HomeWork](https://github.com/0age/HomeWork), maybe give us some new idea.

0age commented:
One thing I’d mention here is that this block of code is suboptimal:
```
            if iszero(extcodesize(_addr)) {
                revert(0, 0)
            }
        }
```
It prevents using the factory to deploy contracts that don’t have runtime code, and also does not return a revert message if deployment fails.

Instead, how about this:
```
      // Pass along revert message if the contract did not deploy successfully.
      if iszero(_addr) {
        returndatacopy(0, 0, returndatasize)
        revert(0, returndatasize)
      }
```

And thanks @skyge for the nod to HomeWork - it is a solution for the case of using fixed bytecode to deploy arbitrary contracts with constructor support. (Maybe premature to include in this create2 factory, but it’s already deployed to mainnet and all the testnets if anyone wants to use it!)

nventuro commented:
Ah, that is a great point @0age, thanks!

@cwhinfrey iirc you mentioned you were planning on tackling this, is that still the case? No pressure, just want to avoid having people do duplicated work :)

AugustoL commented:
Going to add the changes to return address on deploy next week.
I also think there is no need for events. Wdyt @nventuro?

About compute address functions, they are still very useful to have it in the library, there are cases where it is needed to be used on chain.

cwhinfrey commented:
@nventuro Started working on changing the lib based on this conversation but got a bit distracted last week. @AugustoL here's what I have for the lib though. It's a simplified version of what you have in this PR but mostly the same. Main differences are:
- Event was removed
- `computeAddress` function that includes the deployer was removed
- @0age's suggestion was added
Threw it in a gist here: https://gist.github.com/cwhinfrey/1d234fbf00ef71e13b29b1d843068548
@AugustoL Let me know what you think. Happy to PR it to your branch or feel free to just add it in yourself if this is on the right track.

Skyge commented:
> I was thinking in maintaining the two methods, just make one call the other if you dont pass the first param (deployer address). The code inside the two function are same.

@ngmachado I think Solidity does not  have the default argument, so that we can write some override function.

ngmachado commented:
Hi, 

@Skyge, i dont think that solidity as default parameters as for example C#, but we can have the same behavior if we encapsulate the methods and explicit pass the default value. 

For clarity see : https://gist.github.com/ngmachado/1884b97875654d3aab240c5d1ec9cae9

@cwhinfrey as proposed the removal of the method with a deployer param. I'm not in favor of this solution. This library should be pretty agnostic about the usage, and locking the 'consumer' to only deploy new contracts with the callers address creates some unnecessary restrain.  

Naive example: 
One user can create a Proxy to delegate the creation of new contract but want to control the address + salt of the address. 

As the topic on salt type, i think that should be bytes32, a user can if we want, to hash the value of the desirable int. This should be make clear in the usage example of the methods. 

As last note @0age make a very good observation about the validation.

Skyge commented:
@ngmachado Oh, I see, the key point is at line [19](https://gist.github.com/ngmachado/1884b97875654d3aab240c5d1ec9cae9#file-override-sol-L19).
Besides, yeah, the salt type should be bytes32.

AugustoL commented:
About the compute address we have three options, keep only computeAddress with deployer arg, keep only computeAddress without deployer arg or keep both.
IMO the library shouldnt make any restriction about the deployer address used for computation, it should be as flexible and lightweight. So my proposal is to keep only one function with the deployer arg, since the deployer address would be always known. Or keep both functions with different names we can have computeAddress and computeAddressWithDeployer.

For now Im going to commit the changes to have both functions with different names. There is no problem in having both functions with the same name since the function signature would be different, but I guess thats against the dev rules in zep repo.

cwhinfrey commented:
@AugustoL Now that you mention it, I see the use case for both. The library may be used by a factory contract to determine the future address of a contract it will deploy or it may be used by a non-factory contract that will want `computeAddressWithDeployer` to determine where a factory will deploy a given contract. I like the way you have it now.

cwhinfrey commented on contracts/utils/Create2.sol:
```diff
@@ -0,0 +1,73 @@
+     */
+    function computeAddressWithDeployer(
+        address deployer, bytes32 salt, bytes memory code
+    ) internal view returns (address) {
```
I think this can be restricted to `pure`

ngmachado commented:
I'm in favor of both methods, but with the same name. The util provided is the same.
Maybe i'm missing something about solidity and the overrides.

I think that the parameter name is clear enough.

AugustoL commented:
@ngmachado me too. @nventuro and @cwhinfrey ?

cwhinfrey commented:
I don't see any issue with having the same name. Especially since it's a library so it shouldn't ever be called from javascript which is where I feel like it gets difficult to deal with functions that share names.

nventuro commented on contracts/utils/Create2.sol:
```diff
@@ -0,0 +1,73 @@
+     * @return the computed address of the smart contract.
+     */
+    function computeAddressWithDeployer(
+        address deployer, bytes32 salt, bytes memory code
```
```suggestion
        bytes32 salt, bytes memory code, address deployer
```

Nitpicky, but I think this order makes it more consistent with `computeAddress`.

nventuro commented on test/helpers/create2.js:
```diff
@@ -0,0 +1,29 @@
+};
+
+// encodes parameter to pass as contract argument
+const encodeParam = function (dataType, data) {
```
I'd remove `encodeParam` and `isContract` if we won't be needing them for the tests.

General-use test helpers should be added to [openzeppelin-test-helpers](https://github.com/OpenZeppelin/openzeppelin-test-helpers).

nventuro commented:
@AugustoL looks like CI is failing due to outdated usage of the test helpers. I'd suggest merging with the latest `master`, and following the steps in the [migration guide for the helpers](https://github.com/OpenZeppelin/openzeppelin-test-helpers/blob/master/CHANGELOG.md#040-2019-05-23) (i.e. replace `shouldFail` with `expectRevert`).

Skyge commented:
Maybe you can squash this PR down to one commit and rebase.

nventuro commented:
Hm, it looks like our coverage tool doesn't properly support `if` statements in assembly. I'll check what we can do about it.

AugustoL commented:
@nventuro fixed it by skiping Create2.sol in coverage.

nventuro commented:
It looks like the latest version of `solidity-coverage` will be able to support this, but we can't yet upgrade due to https://github.com/sc-forks/solidity-coverage/issues/378. Once that's fixed, we should upgrade and merge this.

nventuro commented:
@frangio I've upgraded `solidity-coverage` to a version that properly supports inline assembly, and updated the documentation to the new format, this should be ready to be merged now.

frangio commented on contracts/utils/Create2.sol:
```diff
@@ -0,0 +1,53 @@
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            addr := create2(0, add(bytecode, 0x20), mload(bytecode), salt)
+            if iszero(extcodesize(addr)) {
```
This check here can give false positives. You can use `create2` to deploy a selfdestructing contract, for example to run a series of transactions with a  deterministic address as the sender. (Interestingly you would be able to reuse this in a new transaction, I think.) In that case, `extcodesize` will return zero even though the transaction has succeeded, and we would revert unnecessarily.

It seems to me we should be able to check `iszero(addr)`? It's not clear from the `CREATE2` EIP whether it returns zero on error, and I can't find the opcode in the yellow paper, but `CREATE` seems to have this behavior.

Alternatively we can use `extcodehash`.

frangio commented on contracts/utils/Create2.sol:
```diff
@@ -0,0 +1,53 @@
+     * a contract cannot be deployed twice using the same salt.
+     */
+    function deploy(bytes32 salt, bytes memory bytecode) internal returns (address) {
+        return _deploy(salt, bytecode);
```
Why have the separate private `_deploy` function? It doesn't seem to add any value. I would inline it.

nventuro replied:
This may be a remnant from an previous iteration, we can get rid of it.

frangio commented on contracts/mocks/Create2Impl.sol:
```diff
@@ -0,0 +1,17 @@
+
+import "../utils/Create2.sol";
+
+contract Create2Impl {
```
I would like to see a test that deploys with `Create2` from within Solidity. It would look like `Create2.deploy(salt, type(ERC20).creationcode)`.

nventuro replied:
How is this different from the `deploy(bytes32 salt, bytes memory code) public` test?

frangio commented on contracts/utils/Create2.sol:
```diff
@@ -0,0 +1,53 @@
+ * See the https://eips.ethereum.org/EIPS/eip-1014#motivation[EIP] for more
+ * information.
+ */
+library Create2 {
```
I'm not a fan of this name. Do we have other options?

nventuro replied:
Usage will be `Create2.deploy()`. I also dislike the name, but I'm not sure how we can more succintly convey the underlying idea.
